python_modules/dagster/dagster/_daemon/backfill.py

```diff
@@ -1,14 +1,115 @@
+from dagster._core.instance import DagsterInstance
+from dagster._core.scheduler.instigation import (
+    InstigatorTick,
+    TickData,
+    TickStatus,
+)
 from dagster._core.workspace.context import IWorkspaceProcessContext
 from dagster._daemon.utils import DaemonErrorCapture
 from dagster._utils.error import SerializableErrorInfo
 
+_BACKFILL_ORIGIN_ID = "backfill_daemon_origin"
+
+
+class BackfillLaunchContext:
+    def __init__(
+        self,
+        tick: InstigatorTick,
+        backfill_id: str,
+        instance: DagsterInstance,
+        logger: logging.Logger,
+        tick_retention_settings,
+    ):
+        self._tick = tick
+        self._backfill_id = backfill_id
+        self._logger = logger
+        self._instance = instance
+
+        self._purge_settings = defaultdict(set)
+        for status, day_offset in tick_retention_settings.items():
+            self._purge_settings[day_offset].add(status)
+
+    @property
+    def status(self) -> TickStatus:
+        return self._tick.status
+
+    @property
+    def tick(self) -> InstigatorTick:
+        return self._tick
+
+    @property
+    def logger(self) -> logging.Logger:
+        return self._logger
+
+    def add_run_info(self, run_id=None):
+        self._tick = self._tick.with_run_info(run_id)
+
+    def set_run_requests(
+        self,
+        run_requests: Sequence[RunRequest],
+        reserved_run_ids: Optional[Sequence[str]],
+    ):
+        self._tick = self._tick.with_run_requests(run_requests, reserved_run_ids=reserved_run_ids)
+        return self._tick
+
+    def update_state(self, status: TickStatus, **kwargs: object):
+        self._tick = self._tick.with_status(status=status, **kwargs)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(
+        self,
+        exception_type: Type[BaseException],
+        exception_value: Exception,
+        traceback: TracebackType,
+    ) -> None:
+        if exception_value and isinstance(exception_value, KeyboardInterrupt):
+            return
+
+        # Log the error if the failure wasn't an interrupt or the daemon generator stopping
+        if exception_value and not isinstance(exception_value, GeneratorExit):
+            error_info = DaemonErrorCapture.on_exception(exc_info=sys.exc_info())
+            self.update_state(TickStatus.FAILURE, error=error_info)
+
+        check.invariant(
+            self._tick.status != TickStatus.STARTED,
+            "Tick must be in a terminal state when the BackfillLaunchContext is closed",
+        )
+
+        # write the new tick status to the database
+
+        self.write()
+
+        for day_offset, statuses in self._purge_settings.items():
+            if day_offset <= 0:
+                continue
+            self._instance.purge_ticks(
+                _BACKFILL_ORIGIN_ID,
+                self._backfill_id,  # TODO issue w actually deleting backfills bc we stop evaluating after backfill completion, need to figure out
```

Comment:
Need to figure out a retention strategy. The issue is that I want each evaluation of a specific backfill to be considered a tick, rather than have each iteration of the daemon (where all backfills are evaluated) considered a single tick. this means we need to have some identifier for each tick tying it to a backfill. but that makes it hard to delete ticks. 

If backfill ABCD takes 10 minutes to execute, but the retention policy is set to a week, by the time the week has elapsed, we are no longer evaluating backfill ABCD in each iteration, so we will never try to purge the ticks